### ISSUE TITLE
\Phalcon\Mvc\Model overrides change to field in belongsTo relation

### ISSUE THREAD

**Issue by Bagarn** [phalcon/cphalcon#12394]
### Expected and Actual Behavior

> Given the following model setup:

```php
use Phalcon\Mvc\Model;
class Status extends Model
{
    public $id;
    public $name;
}
```
```php
use Phalcon\Mvc\Model;
class Item extends Model
{
    public $id;
    public $name;
    public $status_id;

    public function initialize()
    {
        $this->belongsTo(
            'status_id',
            'Status',
            'id',
            ['alias' => 'status']
        );
    }
}
```
***item***

| id [primary ai] | name | status_id |
| - | - | - |
| 1  | My Item  | 1 |

***status***

| id [primary ai] | name |
| - | - |
| 1  | Status 1  | 
| 2  | Status 2  | 

> When I run this:
```php
$item = Item::findFirstById(1);
echo $item->status->id; // 1

$item->status_id = 2;

$item->save();
echo $item->status->id; // 1
```
> I would expect the **Item** to be updated with status_id 2 after the save. 
But what actually happens is that the save takes the Id from the related **Status** object and sets it to the **Item**'s status_id field.

### Details

* Phalcon version:  3.0.1
* PHP Version: 7.0
* Operating System: Ubuntu 16.04 LTS

**Comment by Jurigag** [comment]
Add echo `$item->status_id` please to your code and check for sure which value is it. As well in database.

The problem here is just save doesn't rechange related properties if there are any i think, so status will be kept old if you previously accessed it. This is because first time you access status it's added as public property to object and then if you access it again whole logic for getting related model is not used. Not sure how to fix it. Maybe if there are defined properties in object just always update them if not using snapshot data or check if it was changed and then update them if using it.

**Comment by Bagarn** [comment]
```php
$item = Item::findFirstById(1);
echo $item->status->id; // 1

$item->status_id = 2;
echo $item->status_id; //2 

$item->save();
echo $item->status->id; // 1
echo $item->status_id; // 1
```

> And the database remains unchanged.
> How about clearing the object from _related when one of the keys in the relationship is changed?

**Comment by Jurigag** [comment]
But this is not only about _related. Property is added to object so it's no longer using magic method.

**Comment by Bagarn** [comment]
You could maybe rename it intenally. like this->prefix_{prop_name} ?

**Comment by Jurigag** [comment]
What you mean intenally ? But then magic method would be fired. If you access one time `$item->status` and then next time status property is only used.

**Comment by Bagarn** [comment]
I mean.
When __set('status', 2) is called, one could perhaps check if 'status' is a key in relation and in that case save it to this->relation_key . {'status'} or to an array like this->relation_keys['status'].

Then in the __get method maybe:
if fetch this->relation_keys[prop_name] { return ....

This way the property will be hidden from the outside.
Maybe a bit hackish, but the best I can think of right now.

**Comment by Jurigag** [comment]
But whole point in setting it as property IS NOT TO CALL FUNCTION which is heavy in php.

**Comment by Bagarn** [comment]
Of course. But it would only be the relationship keys. Not the other properties. 
As I said. Hackish.

**Comment by Jurigag** [comment]
No, not hackish. It has to stay the same as it is now - when accessing it first time - use magic method, when second time - use property in object without any method call. Just on save if there is change in property relation then update property(use snapshot data if possible) and that's it.

**Comment by Bagarn** [comment]
Ok. So. Detect the change in the _preSaveRelationTable somehow?

**Comment by gtux** [comment]
I have the same issue, I want to change the parent of the row but it do remains the same even after setting up the value manually. Also, I want to change it after validate one form that receives the values, I mean after bind the data over the entity.

Maybe is there some method to "suspend" or "erase" the belongsTo relationship dynamically in runtime?

**Comment by laszlovl** [comment]
I believe a simple fix would be to add an extra check at https://github.com/phalcon/cphalcon/blob/master/phalcon/mvc/model.zep#L2738:

```php
/**
 * Don't save the referenced model if the current model no longer refers to it,
 * as the foreign key change in the current model would be overwritten.
 */
if this->{columns} == record->readAttribute(referencedFields) {
```

**Comment by Bagarn** [comment]
I might be wrong, but wouldn't that cause problems when creating a new related object.
like so:
```
$item->status = new Status();
```
